apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/SubscriberClient.cs

```diff
@@ -1342,21 +1377,265 @@ private void StartPush()
+                // Failure may be non-retryable due to exceeded time duration or that failure may have become permanent.
+                if (temporaryFailures.Any())
+                {
+                    var retryableErrors = new ConcurrentBag<TimedId>();
+
+                    // Not all temporaryFailures may need to be retried. See if the id exists in the dictionary and only then retry.
+                    foreach (var retryableTemporaryFailure in temporaryFailures.Where(j => _retryableIds.ContainsKey(j.Id)))
+                    {
+                        retryableErrors.Add(retryableTemporaryFailure);
+                    }
+
+                    if (retryableErrors.Count == 0)
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    // All ids in retryableErrors will have the same backoff.
+                    // Get retryInfo for the first retryableError from the _retryableIds dictionary.
+                    if (!_retryableIds.TryGetValue(retryableErrors.First().Id, out var retryInfo))
+                    {
+                        // Nothing to retry.
+                        return;
+                    }
+
+                    var backoff = s_ackBackoff.NextBackoff(retryInfo.CurrentBackoff ?? TimeSpan.Zero);
+                    // If we have errors to retry and we are not past the specified duration, then retry.
+                    if (DateTime.UtcNow.Subtract(retryInfo.FirstTimeOfFailureInUtc).Add(backoff) <= TimeSpan.FromSeconds(durationInSeconds))
+                    {
+                        // Update backoff of retryableErrors Ids in _retryableIds dictionary.
+                        UpdateBackoff(retryableErrors.Select(j => j.Id), backoff);
+                        // Retry errors after backOff.
+                        Task delayTask = _scheduler.Delay(backoff, _softStopCts.Token);
+                        Add(delayTask, new NextAction(false, () =>
+                        {
+                            // This will add the extendIds to the corresponding queue and start push to retry those ids after delay.
+                            // We don't need recursion as the handling of ack response will take care of retry and _retryableIds
+                            // dictionary will ensure that we don't retry after 10 seconds.
+                            actionToRetry(retryableErrors);
+                            StartPush();
+                        }));
+                    }
+                }
+            }
+
+            private void UpdateBackoff(IEnumerable<string> retryableErrors, TimeSpan? backoff)
+            {
+                foreach (var item in retryableErrors)
+                {
+                    if (_retryableIds.TryGetValue(item, out var info))
+                    {
+                        // _retryableIds is a concurrent dictionary but we are updating the value.
```

Comment:
Yes, but that's fine. I don't think Locked is necessary here - and the fact that we don't lock on the dictionary elsewhere means it probably doesn't help anyway.

Would we ever expect to see this change, really? Perhaps we should just use `_retryableIds[item] = info.WithBackoff(backoff);` where `WithBackoff` is a new method on `RetryableInfo`.

Note that if we *do* keep this current code, we should implement `IEquatable<RetryableInfo>` - I'd rather not do that if we can help it.